torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -3155,170 +3155,171 @@ std::tuple<Tensor, Tensor> linalg_lstsq_backward(
+  auto n = R.size(-1);
+
+  TORCH_CHECK(compute_q, "The derivative of linalg.qr is not implemented when mode='r'. "
+                         "Please use torch.linalg.qr(A, mode='reduced')");
+  TORCH_CHECK(reduced || m <= n, "The QR decomposition is not differentiable when "
+                                 "mode='complete' and nrows > ncols.");
+  Tensor gA;
+  if (gQ.defined()) {
+    if (gR.defined()) {
+      gA = gR.matmul(R.mH()) - Q.mH().matmul(gQ);
     } else {
-      // Q is ... x M x N, Q.T is ... x N x M and grad_Q is ... x M x N
-      Q_term = at::zeros_like(R, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+      gA = -Q.mH().matmul(gQ);
     }
-
-    Tensor M = R_term - Q_term;
-
-    // Compute M = (tril(M) + tril(M).mH()) * 0.5 Identity
-    Tensor M_tril = at::tril(M);
-    M = M_tril + M_tril.mH();
-    M.diagonal(0, -2, -1).mul_(0.5);
-
-    Tensor rhs_term;
-    if (grad_Q.defined()) {
-      rhs_term = grad_Q + at::matmul(Q, M);
-    } else {
-      rhs_term = at::matmul(Q, M);
-    }
-
-    // Compute rhs_term @ R^{-H}
-    Tensor grad_A = at::linalg_solve_triangular(
-        R.transpose(-2, -1).conj(),
-        rhs_term,
-        /*upper=*/false,
-        /*left=*/false,
-        /*unitriangular=*/false);
-
-    return grad_A;
-  };
-
-  auto m = self.size(-2);
-  auto n = self.size(-1);
-
-  TORCH_CHECK(
-      ((m <= n && (!reduced)) || reduced),
-      "The derivative of qr is not implemented when mode='complete' and nrows > ncols.");
-
-  auto grad_Q = grads[0];
-  auto grad_R = grads[1];
-
- if (m >= n) {
-    return square_deep_case_backward(grad_Q, grad_R, self, q, r);
   } else {
-    // For wide (m < n) input matrices A,  partition A = [X|Y] and R = [U|V]
-    // X and U are square full rank matrices. We will partition grads,
-    // grad_R = [grad_U | grad_V] and grad_A = [grad_X | grad_Y].
-    // To obtain grad_X we reuse the gradient formula from the square case.
-    // Formulae: grad_X = square_case_grad(grad_Q_prime, grad_U, Q, U),
-    // where grad_Q_prime = grad_Q + Y @ grad_V^H
-    // and grad_Y = Q @ grad_V.
-    // Then concatenate grads to get grad_A = [grad_X | grad_Y].
-
-    auto Y = self.narrow(-1, m, n - m);
-    auto U = r.narrow(-1, 0, m);
-    Tensor grad_Y, grad_X, grad_V, grad_Q_prime;
-
-    if (grad_R.defined()) {
-      grad_V = grad_R.narrow(-1, m, n - m);
-      // reuse grad_R to store grad_U
-      grad_R = grad_R.narrow(-1, 0, m);
-      // grad_Q_prime starts with the value of Y @ grad_V^H
-      grad_Q_prime = at::matmul(Y, grad_V.mH());
-    } else {
-      // when grad_R is not defined then grad_V and grad_Q_prime
-      // get initialized with zeros
-      grad_V = at::zeros_like(Y, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
-      grad_Q_prime = at::zeros_like(q, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+      gA = gR.matmul(R.mH());
+  }
+  if (m >= n) {
+    const auto syminvadj = [](const Tensor& X) {
+      return X + X.mH() - at::diag_embed(at::real(X.diagonal(0, -2, -1)));
```

Comment:
@soulitzer  If you change this by, say
```c++
auto ret = X + X.mH();
at::real(ret.diagonal(0,-2, -1)).mul_(0.5);
return ret;
```
the composite compliance test breaks.

Reply:
Ok the issue seems to be that when we wrap the output in CompositeCompliantTensor, `_make_wrapper_subclass` ignores the strides if it sees that the size is zero and always sets it to 1.

```
elem = torch.ones(0,).as_strided((0,), (2,), 0)
ret = CompositeCompliantTensor(elem)
print(f"{ret.shape = }, {Tensor.stride(ret) = }, {ret.elem.shape = }, {ret.elem.stride() = }")
```

Prints:
```
ret.shape = torch.Size([0]), Tensor.stride(ret) = (1,), ret.elem.shape = torch.Size([0]), ret.elem.stride() = (2,)
```

Either we fix `make_wrapper_subclass` or update CompositeCompliantTensor to skip the check for strides when size is 0.

cc @zou3519

Reply:
fwiw, once this problem is fixed, I think that a number of `linalg` functions will start being composite compliant. For exmaple the eigenvalue decompositions.